py/bitbox02/bitbox02/communication/bitbox_api_protocol.py

```diff
@@ -270,21 +301,133 @@ def _create_noise_channel(self, noise_config: BitBoxNoiseConfig) -> NoiseConnect
             noise_config.add_device_static_pubkey(remote_static_key)
         return noise
 
-    def close(self) -> None:
-        self._transport.close()
+    def noise_connect(self, noise_config: BitBoxNoiseConfig) -> None:
+        self._noise = self._create_noise_channel(noise_config)
+
+    @abstractmethod
+    def unlock_query(self) -> bytes:
+        """
+        Executes an unlock query.
+        Returns the bytes containing the response status.
+        """
+        ...
+
+
+class BitBoxProtocolV1(BitBoxProtocol):
+    """ BitBox Protocol from firmware V1.0.0 onwards. """
+
+    def unlock_query(self) -> bytes:
+        raise NotImplementedError("unlock_query is not supported in BitBox protocol V1")
+
+    def _encode_noise_request(self, encrypted_msg: bytes) -> bytes:
+        return encrypted_msg
+
+    def _decode_noise_response(self, encrypted_msg: bytes) -> Tuple[bytes, bytes]:
+        """
+        Until V7 of the protocol, we don't encapsulate OP_NOISE_MSG responses.
+        Let's assume that if a response is only one byte long, that means it
+        contains an error.
+        """
+        if len(encrypted_msg) == 1:
```

Comment:
before the introduction of `OP_NOISE_MSG`, a noise response could never be 1 byte, could it? it would be zero in case noise decryption failed:

```
if (noise_cipherstate_decrypt(_recv_cipher, &noise_buffer) != NOISE_ERROR_NONE) {
        return false;
    }
```

Reply:
Fixed in cae3727. Checking for empty message instead.